### ISSUE TITLE
Add support for config_settings in requirements.txt / requirements.in

### ISSUE THREAD

**Issue by tiptenbrink** [astral-sh/uv#3636]
In #1460, the CLI gained support for passing config_settings to the build backend. However, sometimes you want to pass these settings only for specific requirements (or simply you don't want to have to write it in the CLI each time). 

If you use `pip install -r requirements.txt`, this is a valid line inside a requirements.txt:

```
--config-settings editable_mode=strict -e ../lib 
```

However, `uv` will fail with:

```
error: Unexpected '-', expected '-c', '-e', '-r' or the start of a requirement at requirements.txt:3:1
```

Now, I can still get things to work if I pass the `--config-settings editable_mode=strict` to the CLI, but it's quite tedious to always have to do that. Furthermore, maybe I'd not want to pass it for a different requirement, but now it applies to all of them.

Note that the same also doesn't work with `uv pip compile` (when putting the `--config-settings editable_mode=strict` inside a requirements.in) and also not with `uv pip sync`. 

It would be awesome if `uv` could parse the `--config-settings` inside a requirements.txt file.

### Why is this useful?

If you use an editable install of a package built with setuptools, you need to pass either `editable_mode=compat` or `editable_mode=strict` in order for pyright/pylance typing to work for the package (see [this StackOverflow question](https://stackoverflow.com/questions/76213501/python-packages-imported-in-editable-mode-cant-be-resolved-by-pylance-in-vscode)). Supporting this means that it becomes a lot simpler to depend on other local packages without requiring them to be fully factored out as PyPI or Git repository packages (all you need is a very basic pyproject.toml with setuptools as the build backend).

My example setup looks like this:

```
lib/
├─ lib.py
├─ pyproject.toml
app/
├─ requirements.in
├─ app.py
```

Also, thanks for this great project! Using it has been a blast so far!

**Comment by charliermarsh** [comment]
Interesting, I've never seen this -- thanks for raising. Looks like it's https://github.com/pypa/pip/issues/11325 in pip.

**Comment by zanieb** [comment]
Interesting. I wonder if there should be a way to express this in the CLI as well.

**Comment by tiptenbrink** [comment]
Small note: If anyone is running into my specific use case: if you use hatchling instead of setuptools as your build backend, pyright will work by default.

**Comment by Jorricks** [comment]
Also faced with this issue unfortunately; https://github.com/microsoft/LightGBM/pull/6493

**Comment by zanieb** [comment]
Considering some options here...

- `--config-setting <package-name>__<key>=<val>` e.g. `--config-setting foo__editable_mode=strict`
- `--config-setting <package>=<key=val> e.g. `--config-setting foo=editable_mode=strict`
- `--config-setting-includes <package>` only pass config settings to the given packages
- `--config-setting-excludes <package` pass config settings to all packages but the given packages
- `--config-setting-includes <package>=<key>` only pass the given config setting to the given packages
- `--config-setting-excludes <package>=<key>` do not pass the given config setting to the given packages

**Comment by ion-elgreco** [comment]
Any chance this will get some attention soon? We have quite some editable installs which we now install one by one

**Comment by zanieb** [comment]
We need to settle on a design — we're prioritizing some other things right now but if someone wants to push the design forward I can review.

**Comment by notatallshaw** [comment]
I'm honestly +0 on this issue, but I thought it worth mentioning I hit this with using `uv pip sync`, at work we have some internal packages that are built with setuptools, and developed in a multi repo environment, to get IDEs to play nice I have to use `--config-settings editable_mode=compat` to install them. 

Because I can't insert that straight into the requirements file I have to do this slightly awkard to of `uv pip sync` followed by `uv pip install {my_editable_projects} --config-settings editable_mode=compat`.

Maybe I'm missing something obvious though.

**Comment by s-rog** [comment]
Can package config-settings be specified in toml files?
Or is this currently a cli exclusive feature?

**Comment by raffaem** [comment]
Also hit that: https://github.com/HandsOnLLM/Hands-On-Large-Language-Models/issues/62

**Comment by ds-steventondeur** [comment]
I've also been hitting the issue. My use case is a namespace package consisting of several subpackages (each one a wheel). In our dev environment we 'self-install' all subpackages in editable mode via `requirements.xt`. In order to play nice with IDEs (VSCode, PyCharm) we need `--config-settings editable_mode=compat`

I.e. something like

```
mypkg/
   + src/
       + mypkg-core/
       + mypkg-tools/
       + ...
```

and `requirements.txt` containing
```
...
-e ./src/mypkg-core  --config-settings editable_mode=compat
-e ./src/mypkg-tools  --config-settings editable_mode=compat
...
```

I'm working in a risk-averse production environment, where we set up our venv using a self-developed minimal tool that basically creates a venv and then runs `pip install -r requirements.txt`. 
We're planning to transition towards `uv` but must do so via a 'vanilla pip'-compatible route, such that each project may decide to use `uv` or not. This means that - in practice - I have to find a way to run `uv pip` on a requirements file that contains config-settings. If that wouldn't have been the case I would've simply used a `uv.toml` file containing
```
config-settings = { editable_mode = "compat" }
```

but alas...

**Comment by ds-steventondeur** [comment]
> Can package config-settings be specified in toml files? Or is this currently a cli exclusive feature?

@s-rog Yes this can be done for `uv`. I was able to do this via a `uv.toml` file containing at the top level:
```
# Compatible mode so IDEs pick up editable installs correctly.
config-settings = { editable_mode = "compat" }
```

This can also be done via `pyproject.toml`, see https://docs.astral.sh/uv/reference/settings/#config-settings

For vanilla pip however there is no such way, it has to be set via CLI.

**Comment by s-rog** [comment]
> > Can package config-settings be specified in toml files? Or is this currently a cli exclusive feature?
> 
> [@s-rog](https://github.com/s-rog) Yes this can be done for `uv`. I was able to do this via a `uv.toml` file containing at the top level:
> 
> ```
> # Compatible mode so IDEs pick up editable installs correctly.
> config-settings = { editable_mode = "compat" }
> ```
> 
> This can also be done via `pyproject.toml`, see https://docs.astral.sh/uv/reference/settings/#config-settings
> 
> For vanilla pip however there is no such way, it has to be set via CLI.

I'm pretty sure it applies to all packages in the env though.

**Comment by yajo** [comment]
Since [uv supports conflicting dependencies](https://docs.astral.sh/uv/concepts/resolution/#conflicting-dependencies), IMHO the feature should be usable in the dependency specification directly. This way, in a project with conflicting dependencies, each one could be specified differently:

```toml 
[project]
name = "myproject"
version = "0.1.0"
description = "Add your description here"

[project.optional-dependencies]
extra1 = [
    "example --config-setting compat=full",
]
extra2 = [
    "example --config-setting compat=none",
]

[tool.uv]
conflicts = [
    [
      { extra = "extra1" },
      { extra = "extra2" },
    ],
]

```
